ghcide/src/Development/IDE/Core/Debouncer.hs

```diff
@@ -31,28 +31,24 @@ newtype Debouncer k = Debouncer { registerEvent :: Seconds -> k -> IO () -> IO (
 
 -- | Debouncer used in the IDE that delays events as expected.
 newAsyncDebouncer :: (Eq k, Hashable k) => IO (Debouncer k)
-newAsyncDebouncer = Debouncer . asyncRegisterEvent <$> newVar Map.empty
+newAsyncDebouncer = Debouncer . asyncRegisterEvent <$> STM.newIO
 
 -- | Register an event that will fire after the given delay if no other event
 -- for the same key gets registered until then.
---
--- If there is a pending event for the same key, the pending event will be killed.
--- Events are run unmasked so it is up to the user of `registerEvent`
--- to mask if required.
-asyncRegisterEvent :: (Eq k, Hashable k) => Var (HashMap k (Async ())) -> Seconds -> k -> IO () -> IO ()
-asyncRegisterEvent d 0 k fire = do
-    join $ modifyVar d $ \m -> do
-        (cancel, !m') <- evaluate $ Map.alterF (\prev -> (traverse_ cancel prev, Nothing)) k m
-        return (m', cancel)
-    fire
-asyncRegisterEvent d delay k fire = mask_ $ do
-    a <- asyncWithUnmask $ \unmask -> unmask $ do
-        sleep delay
-        fire
-        modifyVar_ d (evaluate . Map.delete k)
-    join $ modifyVar d $ \m -> do
-        (cancel, !m') <- evaluate $ Map.alterF (\prev -> (traverse_ cancel prev, Just a)) k m
-        return (m', cancel)
+asyncRegisterEvent :: (Eq k, Hashable k) => STM.Map k (TVar (Seconds, IO())) -> Seconds -> k -> IO () -> IO ()
+asyncRegisterEvent d delay k fire = join $ atomically $ do
+    prev <- STM.lookup k d
+    case prev of
+        Just v -> writeTVar v (delay, fire) >> return (pure ())
+        Nothing -> return $ do
+            var <- newTVarIO (delay, fire)
```

Comment:
Shouldn't the creation and insertion of this TVar be in the STM transaction, rather than the IO action? Because otherwise we could have two threads hitting the path in line 43 and installing two delayed action threads. I'd have thought we want to either update the TVar or insert it, atomically.

Reply:
Something like
```
do
   var <- newTVar (delay, fire)
   STM.insert var k d
   return $ async $ do ...
```

Reply:
Yes, you are right. Thanks!